pkg/machine/fedora.go

```diff
@@ -0,0 +1,122 @@
+// +build amd64 arm64
+
+package machine
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"regexp"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	githubURL = "http://github.com/fedora-cloud/docker-brew-fedora/"
+)
+
+type FedoraDownload struct {
+	Download
+}
+
+func NewFedoraDownloader(vmType, vmName, releaseStream string) (DistributionDownload, error) {
+	imageName, downloadURL, size, err := getFedoraDownload(releaseStream)
+	if err != nil {
+		return nil, err
+	}
+
+	dataDir, err := GetDataDir(vmType)
+	if err != nil {
+		return nil, err
+	}
+
+	f := FedoraDownload{
+		Download: Download{
+			Arch:      getFcosArch(),
+			Artifact:  artifact,
+			Format:    Format,
+			ImageName: imageName,
+			LocalPath: filepath.Join(dataDir, imageName),
+			URL:       downloadURL,
+			VMName:    vmName,
+			Size:      size,
+		},
+	}
+	f.Download.LocalUncompressedFile = f.getLocalUncompressedName()
+	return f, nil
+}
+
+func (f FedoraDownload) Get() *Download {
+	return &f.Download
+}
+
+func (f FedoraDownload) HasUsableCache() (bool, error) {
+	info, err := os.Stat(f.LocalPath)
+	if err != nil {
+		return false, nil
+	}
+	return info.Size() == f.Size, nil
+}
+
+func truncRead(url string) ([]byte, error) {
+	resp, err := http.Get(url)
+	if err != nil {
+		return nil, err
+	}
+
+	defer func() {
+		if err := resp.Body.Close(); err != nil {
+			logrus.Error(err)
+		}
+	}()
+
+	body, err := ioutil.ReadAll(io.LimitReader(resp.Body, 10*1024*1024))
+	if err != nil {
+		return nil, err
+	}
+
+	_, _ = io.Copy(io.Discard, resp.Body)
+
+	return body, nil
+}
+
+func getFedoraDownload(releaseStream string) (string, *url.URL, int64, error) {
+	dirURL := githubURL + "tree/" + releaseStream + "/" + getFcosArch() + "/"
```

Comment:
nit: do we need `url.Parse` and `base.ResolveReference(u)` to be safe here.

Reply:
The image-path parameter that this data comes from supports arbitrary URLs and is specified by the user so I don't think there is a vector to exploit in this scenario.